gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -238,15 +257,32 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
       }
 
       out.beginObject();
-      try {
-        for (BoundField boundField : boundFields.values()) {
-          if (boundField.writeField(value)) {
-            out.name(boundField.name);
-            boundField.write(out, value);
+
+      boolean writeContents = true;
+      if (useJsog) {
+        JsogRegistry jsogContext = JsogRegistry.get();
+        String id = jsogContext.geId(value);
+        if (id != null) { //Object already registered, use @ref
+          out.name(JsogRegistry.JSOG_FIELD_REF);
+          out.value(id);
+          writeContents = false;
+        } else {
+          id = jsogContext.register(value);
+          out.name(JsogRegistry.JSOG_FIELD_ID);
+          out.value(id);
```

Comment:
Maybe also have to verify that object does not write a `@ref` as well, similar to how `getBoundFields` prevents duplicate field names